source/ice_api.c

```diff
@@ -0,0 +1,1676 @@
+                                 bool checkPort );
+
+static bool Ice_IsSameIpAddress( StunAttributeAddress_t * pAddr1,
+                                 StunAttributeAddress_t * pAddr2,
+                                 bool checkPort );
+
+static bool Ice_FindCandidateFromIp( IceAgent_t * pIceAgent,
+                                     IceIPAddress_t iceIpAddress,
+                                     bool isRemote,
+                                     IceCandidate_t ** ppCandidate );
+
+static void Ice_TransactionIdStoreRemove( TransactionIdStore_t * pTransactionIdStore,
+                                          uint8_t * pTransactionId );
+
+static bool Ice_TransactionIdStoreHasId( TransactionIdStore_t * pTransactionIdStore,
+                                         uint8_t * pTransactionId );
+
+static void Ice_TransactionIdStoreInsert( TransactionIdStore_t * pTransactionIdStore,
+                                          uint8_t * pTransactionId );
+
+static IceResult_t Ice_CreateTransactionIdStore( uint32_t maxIdCount,
+                                                 TransactionIdStore_t * pTransactionIdStore );
+
+static uint64_t Ice_ComputeCandidatePairPriority( IceCandidatePair_t * pIceCandidatePair,
+                                                  uint32_t isLocalControlling );
+
+static uint32_t Ice_ComputeCandidatePriority( IceCandidate_t * pIceCandidate );
+
+static void Ice_InsertCandidatePair( IceAgent_t * pIceAgent,
+                                     IceCandidatePair_t * pIceCandidatePair,
+                                     int iceCandidatePairCount );
+
+static bool Ice_FindCandidatePairWithLocalAndRemoteAddr( IceAgent_t * pIceAgent,
+                                                         IceIPAddress_t * pSrcAddr,
+                                                         IceIPAddress_t * pRemoteAddr,
+                                                         IceCandidatePair_t ** ppCandidatePair );
+
+static IceResult_t Ice_CheckPeerReflexiveCandidate( IceAgent_t * pIceAgent,
+                                                    IceIPAddress_t * pIpAddr,
+                                                    IceCandidate_t * pLocalCandidate,
+                                                    uint32_t priority,
+                                                    bool isRemote );
+
+/*-------------------------------------------------------------------------------------------------*/
+
+/* Ice_insertCandidatePair : This API is called internally to insert candidate paits based on decreasing priority. */
+
+static void Ice_InsertCandidatePair( IceAgent_t * pIceAgent,
+                                     IceCandidatePair_t * pIceCandidatePair,
+                                     int iceCandidatePairCount )
+{
+    int i;
+    int pivot = -1;
+    IceCandidatePair_t pivotCandidatePair = *pIceCandidatePair;
+
+    if( ( pIceAgent != NULL ) && ( pIceCandidatePair != NULL ) )
```

Comment:
Else we might need to return some error, not sure why the return type is void.

Reply:
updated.